File: praxis1.rb
#!/usr/bin/env ruby
# http://programmingpraxis.com/2014/02/18/two-interview-questions/

# question 1
# given n, find smallest three digit number such that the product of its digits is n

def find_smallest_three_digit_product(n,t=100)
  digit_product = -> i { i.to_s.split(//).map { |e| e.to_i }.inject(:*) }
  t += 1 until digit_product[t] == n
  return t
end

p find_smallest_three_digit_product(100)

# question 2
def compare_arrays(a,b)
  b.sort - a.sort
end

one = [4,7,9,6,1,3,8] 
two = [3,4,1,8,2,9,5,7,6]
p compare_arrays(one,two)

File: praxis1.swift
// http://programmingpraxis.com/2014/02/18/two-interview-questions/

func digitProduct(n: Int, acc: Int = 1) -> (Int, Int) {
  if n < 10 {
    return (n, n * acc)
  } else {
    let newAcc = acc * (n % 10)
    let newN = n / 10
    return digitProduct(newN, acc: newAcc)
  }
}

func findSmallestThreeDigitProduct(n: Int) -> Int {
  var counter = 100
  while n != digitProduct(counter).1 {
    counter++
  }
  return counter
}

println(findSmallestThreeDigitProduct(100))

func compareArrays(a: [Int], b: [Int]) -> [Int] {
  var disjunction = [Int]()
  for element in b {
    if !contains(a,element) {
      disjunction.append(element)
    }
  }
  return disjunction
}

let one = [4,7,9,6,1,3,8] 
let two = [3,4,1,8,2,9,5,7,6]
println(compareArrays(one,two))
